# EECS 4412 Phase 3
#### Param Balaganeshan<br>215686694<br>param11@my.yorku.ca

## Pre-requisite

This project phase requires the data you collected in phase 2. Use the csv file(s) that you have submitted for Task 0 in the phase 2.

## Task 0: Pre-processing for association mining

#### 1. Combine the data of 10 companies into one csv file by adding a new dimension “Company”.
Refer to Phase 2 of Task 1 where this was already completed, the data was merged and added a column "Stock Name". Here we will simply change the column "Stock Name" to "Company".

In [2]:
import pandas as pd
import numpy as np

all_data = pd.read_csv('data/merged_data.csv')
all_data = all_data.rename(columns={'Stock Name': 'Company'})
# Drop unneeded index column
all_data = all_data.drop('Unnamed: 0', axis = 1)
all_data

Date        Open        High         Low       Close   Adj Close  \
0     2022-06-29  137.460007  140.669998  136.669998  139.229996  138.403564   
1     2022-06-30  137.250000  138.369995  133.770004  136.720001  135.908478   
2     2022-07-01  136.039993  139.039993  135.660004  138.929993  138.105347   
3     2022-07-05  137.770004  141.610001  136.929993  141.559998  140.719727   
4     2022-07-06  141.350006  144.119995  141.080002  142.919998  142.071655   
...          ...         ...         ...         ...         ...         ...   
2514  2023-06-22  250.770004  265.000000  248.250000  264.609985  264.609985   
2515  2023-06-23  259.290009  262.450012  252.800003  256.600006  256.600006   
2516  2023-06-26  250.070007  258.369995  240.699997  241.050003  241.050003   
2517  2023-06-27  243.240005  250.389999  240.850006  250.210007  250.210007   
2518  2023-06-28  249.699997  259.880005  248.889999  256.239990  256.239990   

         Volume Company  
0      66242400    AAPL  
1      98964500    AAPL  
2      71051600    AAPL  
3      73353800    AAPL  
4      74064300    AAPL  
...         ...     ...  
2514  166875900    TSLA  
2515  176584100    TSLA  
2516  179990600    TSLA  
2517  164968200    TSLA  
2518  159148200    TSLA  

[2519 rows x 8 columns]

#### 2. Only keep the dimensions: Date, Open, Close, Volume, Company; drop the other dimensions.

We need to drop "High", "Low" and "Adj Close"

In [3]:
all_data = all_data.drop(['High', 'Low', 'Adj Close'], axis=1)
all_data

Date        Open       Close     Volume Company
0     2022-06-29  137.460007  139.229996   66242400    AAPL
1     2022-06-30  137.250000  136.720001   98964500    AAPL
2     2022-07-01  136.039993  138.929993   71051600    AAPL
3     2022-07-05  137.770004  141.559998   73353800    AAPL
4     2022-07-06  141.350006  142.919998   74064300    AAPL
...          ...         ...         ...        ...     ...
2514  2023-06-22  250.770004  264.609985  166875900    TSLA
2515  2023-06-23  259.290009  256.600006  176584100    TSLA
2516  2023-06-26  250.070007  241.050003  179990600    TSLA
2517  2023-06-27  243.240005  250.210007  164968200    TSLA
2518  2023-06-28  249.699997  256.239990  159148200    TSLA

[2519 rows x 5 columns]

#### 3. Sort the data on date. Now you will have 10 rows of data per day where each row represents a different company.

In [4]:
all_data = all_data.sort_values(by=['Date', 'Company'])
all_data

Date        Open       Close     Volume Company
0     2022-06-29  137.460007  139.229996   66242400    AAPL
252   2022-06-29   79.550003   77.989998  104140900     AMD
504   2022-06-29  107.379997  108.919998   66375300    AMZN
756   2022-06-29  112.148499  112.256500   18628000    GOOG
1008  2022-06-29  142.740005  140.710007    4161500     IBM
...          ...         ...         ...        ...     ...
1259  2023-06-29  131.750000  133.854996    2559865     IBM
1511  2023-06-29   33.340000   32.904999   27390860    INTC
1763  2023-06-29    6.475000    6.855000   61880326    LCID
2015  2023-06-29  284.500000  281.440002   11050361    META
2267  2023-06-29  334.709991  334.200012   11173217    MSFT

[2519 rows x 5 columns]

#### 4. Add 10 new dimensions to the data, 1 for each company. You can name these dimensions same as the company name.

In [5]:
transformed_data = pd.DataFrame(0, index = range(252), columns=[ 'Date',
                                                        'AAPL', 
                                                         'AMD', 
                                                        'AMZN',
                                                        'GOOG',
                                                         'IBM',
                                                        'INTC',
                                                        'LCID',
                                                        'META',
                                                        'MSFT',
                                                        'TSLA'])
transformed_data

Date  AAPL  AMD  AMZN  GOOG  IBM  INTC  LCID  META  MSFT  TSLA
0       0     0    0     0     0    0     0     0     0     0     0
1       0     0    0     0     0    0     0     0     0     0     0
2       0     0    0     0     0    0     0     0     0     0     0
3       0     0    0     0     0    0     0     0     0     0     0
4       0     0    0     0     0    0     0     0     0     0     0
..    ...   ...  ...   ...   ...  ...   ...   ...   ...   ...   ...
247     0     0    0     0     0    0     0     0     0     0     0
248     0     0    0     0     0    0     0     0     0     0     0
249     0     0    0     0     0    0     0     0     0     0     0
250     0     0    0     0     0    0     0     0     0     0     0
251     0     0    0     0     0    0     0     0     0     0     0

[252 rows x 11 columns]

#### 5. Merge the 10 rows of each day in the following manner:
a. If the price of a company has gone up i.e., Close-Open is positive then copy the volume value in the dimension of that company, the one that you added in step 4. Otherwise add 0. <br>
b. In the end you will have only 1 row per date and the 10 dimensions added in step 4 will have either 0 or volume-value indicating whether the price of this stock has gone down or up (in this case how many units sold). <br>
c. Drop the Open, Close, Volume and Company dimensions.

In [6]:
# Creating copy of dataframe to store transformed data
all_data_copy = all_data.copy()

for index, row in all_data_copy.iterrows():
    if row['Close'] - row['Open'] <= 0:
        all_data_copy.at[index, 'Volume'] = 0

all_data_copy = all_data_copy.drop(columns = ['Open', 'Close'], axis = 1)
all_data_copy = all_data_copy.reset_index(drop=True)
all_data_copy

Date    Volume Company
0     2022-06-29  66242400    AAPL
1     2022-06-29         0     AMD
2     2022-06-29  66375300    AMZN
3     2022-06-29  18628000    GOOG
4     2022-06-29         0     IBM
...          ...       ...     ...
2514  2023-06-29   2559865     IBM
2515  2023-06-29         0    INTC
2516  2023-06-29  61880326    LCID
2517  2023-06-29         0    META
2518  2023-06-29         0    MSFT

[2519 rows x 3 columns]

In [7]:
import math

# Fully transform data
for index, row in all_data_copy.iterrows():
    row_num = math.floor(index/10)
    col_num = (index % 10) + 1
    transformed_data.iloc[row_num, col_num] = row['Volume']

for x in range(252):
    transformed_data.iloc[x, 0] = all_data_copy.iloc[x*10, 0]

transformed_data

Date      AAPL       AMD      AMZN      GOOG      IBM      INTC  \
0    2022-06-29  66242400         0  66375300  18628000        0         0   
1    2022-06-30         0         0         0         0  4878000  35141000   
2    2022-07-01  71051600         0  73021200  31028000  4011900         0   
3    2022-07-05  73353800  95589000  76583700  36398000        0  33215000   
4    2022-07-06  74064300  85458900  66958900  28852000  3237000  26258300   
..          ...       ...       ...       ...       ...      ...       ...   
247  2023-06-23  53079300  73545600  71855200  29542900        0  46199200   
248  2023-06-26         0         0         0         0  4845600  38952400   
249  2023-06-27  50730800  59381600  46801000  27221700  3219900  33038900   
250  2023-06-28  51133200  67295100  51695900  19731800        0         0   
251  2023-06-29  32866759         0         0         0  2559865         0   

         LCID      META      MSFT       TSLA  
0           0  28595200  20069800          0  
1           0         0         0          0  
2           0         0  22837700   74460300  
3    14868300  28618600  22941000   84581100  
4    21390600  23085900  23824400   71853600  
..        ...       ...       ...        ...  
247         0  50988400  23084700          0  
248         0         0         0          0  
249  74181400  26108300  24354100  164968200  
250  68049600  16599700  20011000  159148200  
251  61880326         0         0          0  

[252 rows x 11 columns]

This generated the transformed data that we are going to use for further analysis in the next tasks. 

#### 6. Final Check
a. You have Date and 10 company dimensions, 11 in total. <br>
b. Number of rows is equal to the number of days the market was open during the year

In [8]:
transformed_data

Date      AAPL       AMD      AMZN      GOOG      IBM      INTC  \
0    2022-06-29  66242400         0  66375300  18628000        0         0   
1    2022-06-30         0         0         0         0  4878000  35141000   
2    2022-07-01  71051600         0  73021200  31028000  4011900         0   
3    2022-07-05  73353800  95589000  76583700  36398000        0  33215000   
4    2022-07-06  74064300  85458900  66958900  28852000  3237000  26258300   
..          ...       ...       ...       ...       ...      ...       ...   
247  2023-06-23  53079300  73545600  71855200  29542900        0  46199200   
248  2023-06-26         0         0         0         0  4845600  38952400   
249  2023-06-27  50730800  59381600  46801000  27221700  3219900  33038900   
250  2023-06-28  51133200  67295100  51695900  19731800        0         0   
251  2023-06-29  32866759         0         0         0  2559865         0   

         LCID      META      MSFT       TSLA  
0           0  28595200  20069800          0  
1           0         0         0          0  
2           0         0  22837700   74460300  
3    14868300  28618600  22941000   84581100  
4    21390600  23085900  23824400   71853600  
..        ...       ...       ...        ...  
247         0  50988400  23084700          0  
248         0         0         0          0  
249  74181400  26108300  24354100  164968200  
250  68049600  16599700  20011000  159148200  
251  61880326         0         0          0  

[252 rows x 11 columns]

As seen above, the generated data now has "Date" and 10 company dimensions. We also have 252 rows which is the amount of days the market was open during 2022-06-29 to 2023-06-29. 

In [9]:
transformed_data.to_csv('data/transformed_data.csv')

## Task 1: Association Analysis
Our objective is to identify which stocks/companies go up together. Following this objective, answer the questions below.

#### Find the 10 most frequent item-sets. The length/size of these item-sets must be at least 2. Compute the support for these item-sets.

The following libraries will be the tools used for finding frequent item-sets and association rules.

In [10]:
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

Now the data needs to be transformed to use the Apriori algorithm. The following steps will encode the data so that the mlxtend library can use the apriori algorithm. The encoding simply will transform the data where any positive values will become value "True/1". If zero, it will remain "False/0". We will also drop date as we are simply finding which stocks move up in price regardless of date. Once all positive values are transformed to ones, now the apriori algorithm can take the data as an input can detect which stocks move up in price with each other.

In [31]:
tr = transformed_data.copy()
tr = tr.drop(['Date'], axis=1)
for row_index, row in tr.iterrows():
    for col_index, value in row.items():
        if value > 0:
            tr.loc[row_index, col_index] = 1
tr

AAPL  AMD  AMZN  GOOG  IBM  INTC  LCID  META  MSFT  TSLA
0       1    0     1     1    0     0     0     1     1     0
1       0    0     0     0    1     1     0     0     0     0
2       1    0     1     1    1     0     0     0     1     1
3       1    1     1     1    0     1     1     1     1     1
4       1    1     1     1    1     1     1     1     1     1
..    ...  ...   ...   ...  ...   ...   ...   ...   ...   ...
247     1    1     1     1    0     1     0     1     1     0
248     0    0     0     0    1     1     0     0     0     0
249     1    1     1     1    1     1     1     1     1     1
250     1    1     1     1    0     0     1     1     1     1
251     1    0     0     0    1     0     1     0     0     0

[252 rows x 10 columns]

Now that the data has been transformed, Apriori can be applied to the data.

In [82]:
frequent_itemsets = apriori(tr, min_support=0.4, use_colnames=True)
frequent_itemsets = frequent_itemsets.sort_values(by='support', ascending=False)
frequent_itemsets = frequent_itemsets.iloc[:20]
frequent_itemsets

C:\Users\param\anaconda3\lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support      itemsets
0   0.583333        (AAPL)
7   0.567460        (META)
3   0.535714        (GOOG)
8   0.531746        (MSFT)
1   0.527778         (AMD)
4   0.527778         (IBM)
5   0.515873        (INTC)
9   0.515873        (TSLA)
2   0.507937        (AMZN)
6   0.484127        (LCID)
14  0.448413  (AAPL, MSFT)
12  0.448413  (AAPL, GOOG)
13  0.440476  (AAPL, META)
11  0.436508  (AAPL, AMZN)
21  0.424603  (GOOG, META)
23  0.424603  (META, MSFT)
22  0.420635  (GOOG, MSFT)
19  0.416667  (AMZN, META)
20  0.416667  (AMZN, MSFT)
10  0.412698   (AAPL, AMD)

Since the size of these items sets needs to be at size 2, the first 10 rows of the output above will be omitted and the next 10 will be considered the 10 most frequent item-sets.

In [83]:
frequent_itemsets_pruned = frequent_itemsets.iloc[10:20]
frequent_itemsets_pruned

support      itemsets
14  0.448413  (AAPL, MSFT)
12  0.448413  (AAPL, GOOG)
13  0.440476  (AAPL, META)
11  0.436508  (AAPL, AMZN)
21  0.424603  (GOOG, META)
23  0.424603  (META, MSFT)
22  0.420635  (GOOG, MSFT)
19  0.416667  (AMZN, META)
20  0.416667  (AMZN, MSFT)
10  0.412698   (AAPL, AMD)

The support is listed beside the itemsets. As observed they are all around a support of 0.44 to 0.41.

So from the frequent itemset list, we can state that these stocks move up in price with each other.

#### From your 10 most frequent items-sets, design 10 association rules, 5 with most confidence and 5 with least confidence. Specify the confidence for each of these rules.

In [84]:
rules = association_rules(frequent_itemsets, min_threshold=0.001)
rules = rules.sort_values(['confidence'], ascending=False)
rules

antecedents consequents  antecedent support  consequent support   support  \
7       (AMZN)      (AAPL)            0.507937            0.583333  0.436508   
1       (MSFT)      (AAPL)            0.531746            0.583333  0.448413   
3       (GOOG)      (AAPL)            0.535714            0.583333  0.448413   
16      (AMZN)      (MSFT)            0.507937            0.531746  0.416667   
14      (AMZN)      (META)            0.507937            0.567460  0.416667   
11      (MSFT)      (META)            0.531746            0.567460  0.424603   
8       (GOOG)      (META)            0.535714            0.567460  0.424603   
13      (MSFT)      (GOOG)            0.531746            0.535714  0.420635   
12      (GOOG)      (MSFT)            0.535714            0.531746  0.420635   
17      (MSFT)      (AMZN)            0.531746            0.507937  0.416667   
19       (AMD)      (AAPL)            0.527778            0.583333  0.412698   
5       (META)      (AAPL)            0.567460            0.583333  0.440476   
0       (AAPL)      (MSFT)            0.583333            0.531746  0.448413   
2       (AAPL)      (GOOG)            0.583333            0.535714  0.448413   
4       (AAPL)      (META)            0.583333            0.567460  0.440476   
6       (AAPL)      (AMZN)            0.583333            0.507937  0.436508   
9       (META)      (GOOG)            0.567460            0.535714  0.424603   
10      (META)      (MSFT)            0.567460            0.531746  0.424603   
15      (META)      (AMZN)            0.567460            0.507937  0.416667   
18      (AAPL)       (AMD)            0.583333            0.527778  0.412698   

    confidence      lift  leverage  conviction  zhangs_metric  
7     0.859375  1.473214  0.140212    2.962963       0.652786  
1     0.843284  1.445629  0.138228    2.658730       0.658317  
3     0.837037  1.434921  0.135913    2.556818       0.652825  
16    0.820313  1.542677  0.146573    2.605935       0.714900  
14    0.820313  1.445586  0.128433    2.407177       0.626421  
11    0.798507  1.407160  0.122858    2.146678       0.617931  
8     0.792593  1.396737  0.120607    2.085459       0.611790  
13    0.791045  1.476617  0.135771    2.221939       0.689319  
12    0.785185  1.476617  0.135771    2.179803       0.695210  
17    0.783582  1.542677  0.146573    2.273673       0.751251  
19    0.781955  1.340494  0.104828    1.910920       0.537896  
5     0.776224  1.330669  0.109458    1.861979       0.574510  
0     0.768707  1.445629  0.138228    2.024510       0.739823  
2     0.768707  1.434921  0.135913    2.007353       0.727434  
4     0.755102  1.330669  0.109458    1.766204       0.596396  
6     0.748299  1.473214  0.140212    1.954955       0.770909  
9     0.748252  1.396737  0.120607    1.844246       0.656692  
10    0.748252  1.407160  0.122858    1.860009       0.668953  
15    0.734266  1.445586  0.128433    1.851713       0.712626  
18    0.707483  1.340494  0.104828    1.614341       0.609615

5 rules with most confidence is ...

In [89]:
top_5_rules = rules.iloc[:5]
top_5_rules

antecedents consequents  antecedent support  consequent support   support  \
7       (AMZN)      (AAPL)            0.507937            0.583333  0.436508   
1       (MSFT)      (AAPL)            0.531746            0.583333  0.448413   
3       (GOOG)      (AAPL)            0.535714            0.583333  0.448413   
16      (AMZN)      (MSFT)            0.507937            0.531746  0.416667   
14      (AMZN)      (META)            0.507937            0.567460  0.416667   

    confidence      lift  leverage  conviction  zhangs_metric  
7     0.859375  1.473214  0.140212    2.962963       0.652786  
1     0.843284  1.445629  0.138228    2.658730       0.658317  
3     0.837037  1.434921  0.135913    2.556818       0.652825  
16    0.820313  1.542677  0.146573    2.605935       0.714900  
14    0.820313  1.445586  0.128433    2.407177       0.626421

AMZN -> AAPL has confidence of 0.859
<br>
MSFT -> AAPL has confidence of 0.843
<br>
GOOG -> AAPL has confidence of 0.837
<br>
AMZN -> MSFT has confidence of 0.820
<br>
AMZN -> META has confidence of 0.820

5 rules with least confidence is ...

In [91]:
low_5_rules = rules.tail(5)
low_5_rules

antecedents consequents  antecedent support  consequent support   support  \
6       (AAPL)      (AMZN)            0.583333            0.507937  0.436508   
9       (META)      (GOOG)            0.567460            0.535714  0.424603   
10      (META)      (MSFT)            0.567460            0.531746  0.424603   
15      (META)      (AMZN)            0.567460            0.507937  0.416667   
18      (AAPL)       (AMD)            0.583333            0.527778  0.412698   

    confidence      lift  leverage  conviction  zhangs_metric  
6     0.748299  1.473214  0.140212    1.954955       0.770909  
9     0.748252  1.396737  0.120607    1.844246       0.656692  
10    0.748252  1.407160  0.122858    1.860009       0.668953  
15    0.734266  1.445586  0.128433    1.851713       0.712626  
18    0.707483  1.340494  0.104828    1.614341       0.609615

AAPL -> AMZN has confidence of 0.436
<br>
META -> GOOG has confidence of 0.424
<br>
META -> MSFT has confidence of 0.424
<br>
META -> AMZN has confidence of 0.416
<br>
AAPL -> AMD has confidence of 0.412

#### For each of your top five rules, compute any 5 different measures of interest (from Chapter05)

In [93]:
top_5_rules

antecedents consequents  antecedent support  consequent support   support  \
7       (AMZN)      (AAPL)            0.507937            0.583333  0.436508   
1       (MSFT)      (AAPL)            0.531746            0.583333  0.448413   
3       (GOOG)      (AAPL)            0.535714            0.583333  0.448413   
16      (AMZN)      (MSFT)            0.507937            0.531746  0.416667   
14      (AMZN)      (META)            0.507937            0.567460  0.416667   

    confidence      lift  leverage  conviction  zhangs_metric  
7     0.859375  1.473214  0.140212    2.962963       0.652786  
1     0.843284  1.445629  0.138228    2.658730       0.658317  
3     0.837037  1.434921  0.135913    2.556818       0.652825  
16    0.820313  1.542677  0.146573    2.605935       0.714900  
14    0.820313  1.445586  0.128433    2.407177       0.626421

The interest has already been computed by the tool used. Check column "lift" above to see interest values

AMZN -> AAPL has lift/interest of 1.473
<br>
MSFT -> AAPL has lift/interest of 1.445
<br>
GOOG -> AAPL has lift/interest of 1.434
<br>
AMZN -> MSFT has lift/interest of 1.542
<br>
AMZN -> META has lift/interest of 1.445

## Task 2: Clustering Analysis

For this task, data from Phase 2 Task 6 will be used.

In [104]:
all_data_task2 = pd.read_csv('../Phase 2/Data/merged_data.csv')
all_data_task2 = all_data_task2.rename(columns={'Stock Name': 'Company'})
all_data_task2 = all_data_task2.drop('Unnamed: 0', axis = 1)
all_data_task2

Date        Open        High         Low       Close   Adj Close  \
0     2022-06-29  137.460007  140.669998  136.669998  139.229996  138.403564   
1     2022-06-30  137.250000  138.369995  133.770004  136.720001  135.908478   
2     2022-07-01  136.039993  139.039993  135.660004  138.929993  138.105347   
3     2022-07-05  137.770004  141.610001  136.929993  141.559998  140.719727   
4     2022-07-06  141.350006  144.119995  141.080002  142.919998  142.071655   
...          ...         ...         ...         ...         ...         ...   
2514  2023-06-22  250.770004  265.000000  248.250000  264.609985  264.609985   
2515  2023-06-23  259.290009  262.450012  252.800003  256.600006  256.600006   
2516  2023-06-26  250.070007  258.369995  240.699997  241.050003  241.050003   
2517  2023-06-27  243.240005  250.389999  240.850006  250.210007  250.210007   
2518  2023-06-28  249.699997  259.880005  248.889999  256.239990  256.239990   

         Volume Company  
0      66242400    AAPL  
1      98964500    AAPL  
2      71051600    AAPL  
3      73353800    AAPL  
4      74064300    AAPL  
...         ...     ...  
2514  166875900    TSLA  
2515  176584100    TSLA  
2516  179990600    TSLA  
2517  164968200    TSLA  
2518  159148200    TSLA  

[2519 rows x 8 columns]

#### Perform any necessary preprocessing steps. Add explanation to your report.

First preprocessing is neeeded, by binarizing "Date" and "Company".

In [122]:
all_data_task2_copy = all_data_task2.copy()
# Preprocessing for Date
for index, row in all_data_task2_copy.iterrows():
    new_str = row['Date'].replace('-', '')
    all_data_task2_copy['Date'] = all_data_task2_copy['Date'].replace(row['Date'], int(new_str))
all_data_task2_copy

Date        Open        High         Low       Close   Adj Close  \
0     20220629  137.460007  140.669998  136.669998  139.229996  138.403564   
1     20220630  137.250000  138.369995  133.770004  136.720001  135.908478   
2     20220701  136.039993  139.039993  135.660004  138.929993  138.105347   
3     20220705  137.770004  141.610001  136.929993  141.559998  140.719727   
4     20220706  141.350006  144.119995  141.080002  142.919998  142.071655   
...        ...         ...         ...         ...         ...         ...   
2514  20230622  250.770004  265.000000  248.250000  264.609985  264.609985   
2515  20230623  259.290009  262.450012  252.800003  256.600006  256.600006   
2516  20230626  250.070007  258.369995  240.699997  241.050003  241.050003   
2517  20230627  243.240005  250.389999  240.850006  250.210007  250.210007   
2518  20230628  249.699997  259.880005  248.889999  256.239990  256.239990   

         Volume Company  
0      66242400    AAPL  
1      98964500    AAPL  
2      71051600    AAPL  
3      73353800    AAPL  
4      74064300    AAPL  
...         ...     ...  
2514  166875900    TSLA  
2515  176584100    TSLA  
2516  179990600    TSLA  
2517  164968200    TSLA  
2518  159148200    TSLA  

[2519 rows x 8 columns]

In [123]:
#Preprocessing for Company
company_dict = {
    'AAPL': 0,
    'AMD': 1,
    'AMZN': 2,
    'GOOG': 3,
    'IBM': 4,
    'INTC': 5,
    'LCID': 6,
    'META': 7,
    'MSFT': 8,
    'TSLA': 9,
}
for index, row in all_data_task2_copy.iterrows():
    str_int= company_dict[row['Company']]
    all_data_task2_copy['Company'] = all_data_task2_copy['Company'].replace(row['Company'], str_int)
all_data_task2_copy

Date        Open        High         Low       Close   Adj Close  \
0     20220629  137.460007  140.669998  136.669998  139.229996  138.403564   
1     20220630  137.250000  138.369995  133.770004  136.720001  135.908478   
2     20220701  136.039993  139.039993  135.660004  138.929993  138.105347   
3     20220705  137.770004  141.610001  136.929993  141.559998  140.719727   
4     20220706  141.350006  144.119995  141.080002  142.919998  142.071655   
...        ...         ...         ...         ...         ...         ...   
2514  20230622  250.770004  265.000000  248.250000  264.609985  264.609985   
2515  20230623  259.290009  262.450012  252.800003  256.600006  256.600006   
2516  20230626  250.070007  258.369995  240.699997  241.050003  241.050003   
2517  20230627  243.240005  250.389999  240.850006  250.210007  250.210007   
2518  20230628  249.699997  259.880005  248.889999  256.239990  256.239990   

         Volume  Company  
0      66242400        0  
1      98964500        0  
2      71051600        0  
3      73353800        0  
4      74064300        0  
...         ...      ...  
2514  166875900        9  
2515  176584100        9  
2516  179990600        9  
2517  164968200        9  
2518  159148200        9  

[2519 rows x 8 columns]

#### Perform the k-means clustering of your data with k=5 and 7. Which value of k produces the least SSE (sum of squared errors)?

In [124]:
import matplotlib.pyplot as plt
from sklearn.cluster import KMeans

Performing k-means clustering with k=5.

In [132]:
kmeans5 = KMeans(n_clusters=5).fit(all_data_task2_copy)
centroids5 = kmeans5.cluster_centers_
print(centroids5)

C:\Users\param\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
C:\Users\param\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=10.
  warnings.warn(


[[2.02254678e+07 1.29368789e+02 1.30919456e+02 1.28044315e+02
  1.29553599e+02 1.28564725e+02 1.88727230e+07 5.41740675e+00]
 [2.02244167e+07 1.31517254e+02 1.33947058e+02 1.29064015e+02
  1.31559150e+02 1.31388996e+02 7.82532812e+07 2.67004049e+00]
 [2.02285716e+07 1.69456583e+02 1.74975823e+02 1.65097341e+02
  1.70434050e+02 1.70411438e+02 1.86908142e+08 8.18987342e+00]
 [2.02260146e+07 1.13489055e+02 1.15253133e+02 1.11911471e+02
  1.13644223e+02 1.13399184e+02 4.75950904e+07 3.65689150e+00]
 [2.02262499e+07 1.47212488e+02 1.50755048e+02 1.43846908e+02
  1.47567077e+02 1.47489596e+02 1.25969988e+08 5.58695652e+00]]


Performing k-means clustering with k=7.

In [140]:
kmeans7 = KMeans(n_clusters=7).fit(all_data_task2_copy)
centroids7 = kmeans7.cluster_centers_
print(centroids7)

C:\Users\param\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
C:\Users\param\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=10.
  warnings.warn(


[[2.02255128e+07 1.35478789e+02 1.37331661e+02 1.33936498e+02
  1.35695174e+02 1.35287151e+02 2.66133642e+07 5.75845974e+00]
 [2.02280828e+07 1.67738674e+02 1.72050394e+02 1.63699318e+02
  1.68303943e+02 1.68263386e+02 1.55702367e+08 7.25806452e+00]
 [2.02262160e+07 1.15314688e+02 1.17058773e+02 1.13739558e+02
  1.15493690e+02 1.15298331e+02 5.04001373e+07 3.21731123e+00]
 [2.02245016e+07 1.27532950e+02 1.29820125e+02 1.25368725e+02
  1.27697543e+02 1.27519789e+02 7.50537655e+07 2.41549296e+00]
 [2.02286258e+07 1.60981471e+02 1.66912352e+02 1.55846176e+02
  1.61511470e+02 1.61511470e+02 2.12434971e+08 8.64705882e+00]
 [2.02247106e+07 1.45623293e+02 1.49062490e+02 1.42219958e+02
  1.45682201e+02 1.45574068e+02 1.07297457e+08 4.62345679e+00]
 [2.02253336e+07 1.08976322e+02 1.09979522e+02 1.08017408e+02
  1.09045334e+02 1.07004853e+02 8.12903788e+06 4.70792079e+00]]


Next step is to find the Sum of Squared Errors for k=5 and k=7. For this we can use the inertia_ function to obtain the SSE.

For k=5 ...

In [142]:
kmeans5.inertia_

2.900017413748482e+17

For k=7 ...

In [143]:
kmeans7.inertia_

1.439795870759883e+17

As observed, k=7 has a lower sum of squares error, meaning it better models the regression. <br>
ASIDE: You want to have low inertia and low amount of clusters.

#### Consider the same class(es) as in your Phase 2 Task 6. Perform the k-means clustering where k= number of classes in your Phase 2 Task 6.

This means k=10.

In [144]:
kmeans10 = KMeans(n_clusters=10).fit(all_data_task2_copy)
centroids10 = kmeans10.cluster_centers_
print(centroids10)

C:\Users\param\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
C:\Users\param\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=10.
  warnings.warn(


[[2.02235621e+07 1.37872457e+02 1.40740157e+02 1.34845258e+02
  1.37868200e+02 1.37705501e+02 8.86060484e+07 3.10328638e+00]
 [2.02254925e+07 1.42467382e+02 1.44297713e+02 1.40985967e+02
  1.42751627e+02 1.42400623e+02 2.26464492e+07 5.87349398e+00]
 [2.02277890e+07 1.60889453e+02 1.64664863e+02 1.57126174e+02
  1.61036830e+02 1.60986627e+02 1.49049898e+08 6.86885246e+00]
 [2.02267471e+07 1.18327326e+02 1.20066785e+02 1.16780103e+02
  1.18532586e+02 1.18374201e+02 5.22479832e+07 2.73878628e+00]
 [2.02254485e+07 1.08158006e+02 1.09084802e+02 1.07210896e+02
  1.08161322e+02 1.05685853e+02 6.28677631e+06 4.47720365e+00]
 [2.02278485e+07 1.54178261e+02 1.61009130e+02 1.48863913e+02
  1.54733043e+02 1.54733043e+02 2.22595261e+08 8.47826087e+00]
 [2.02256086e+07 1.42011921e+02 1.45625326e+02 1.39027355e+02
  1.42661631e+02 1.42566989e+02 1.16751157e+08 5.07608696e+00]
 [2.02248555e+07 1.26732183e+02 1.28851032e+02 1.24742983e+02
  1.26824783e+02 1.26659391e+02 6.94339484e+07 2.43034056e+00]


Now computing the SSE of k=10.

In [145]:
kmeans10.inertia_

7.109363390846859e+16

As observed, the inertia increased from k=7. This may mean we have overshooted the number of clusters. Recall, you want to have the lowest number of clusters while maintaining low inertia/SSE.
<br><br>
<b> To determine the best number of clusters, use the elbow method. <b>

####  Considering your entire dataset as test data, compute the accuracy of this clustering approach as classification technique.

Aside: Clustering is not a classification tool. Usually not recommended to find a accuracy for clustering techniques as it simply a tool to group similar data. It also does not make sense to use the whole dataset as a test, unless if we are comparing it to another dataset like the one used in phase 2.